crates/bevy_pbr/src/light.rs

```diff
@@ -256,9 +227,243 @@ pub struct DirectionalLightShadowMap {
         #[cfg(not(feature = "webgl"))]
-        return Self { size: 4096 };
+        return Self { size: 2048 };
+    }
+}
+
+/// Controls how cascaded shadow mapping works.
+#[derive(Component, Clone, Debug, Reflect)]
+#[reflect(Component)]
+pub struct CascadeShadowConfig {
+    /// The (positive) distance to the far boundary of each cascade.
+    pub bounds: Vec<f32>,
+    /// The proportion of overlap each cascade has with the previous cascade.
+    pub overlap_proportion: f32,
+}
+
+impl Default for CascadeShadowConfig {
+    fn default() -> Self {
+        Self::new(4, 5.0, 1000.0, 0.2)
+    }
+}
+
+fn calculate_cascade_bounds(
+    num_cascades: usize,
+    nearest_bound: f32,
+    shadow_maximum_distance: f32,
+) -> Vec<f32> {
+    if num_cascades == 1 {
+        return vec![shadow_maximum_distance];
+    }
+    let base = (shadow_maximum_distance / nearest_bound).powf(1.0 / (num_cascades - 1) as f32);
+    (0..num_cascades)
+        .map(|i| nearest_bound * base.powf(i as f32))
+        .collect()
+}
+
+impl CascadeShadowConfig {
+    /// Returns a cascade config for `num_cascades` cascades, with the first cascade
+    /// having far bound `nearest_bound` and the last cascade having far bound `shadow_maximum_distance`.
+    /// In-between cascades will be exponentially spaced.
+    pub fn new(
+        num_cascades: usize,
+        nearest_bound: f32,
+        shadow_maximum_distance: f32,
+        overlap_proportion: f32,
+    ) -> Self {
+        assert!(num_cascades > 0, "num_cascades must be positive");
+        Self {
+            bounds: calculate_cascade_bounds(num_cascades, nearest_bound, shadow_maximum_distance),
+            overlap_proportion,
+        }
+    }
+}
+
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[reflect(Component)]
+pub struct Cascades {
+    /// Cascade configuration, per-view.
+    pub(crate) cascades: HashMap<Entity, Vec<Cascade>>,
+}
+
+#[derive(Clone, Debug, Default, Reflect, FromReflect)]
+pub struct Cascade {
+    /// The transform of the light, i.e. the view to world matrix.
+    pub(crate) view_transform: Mat4,
+    /// The orthographic projection for this cascade.
+    pub(crate) projection: Mat4,
+    /// The view-projection matrix for this cacade, converting world space into light clip space.
+    /// Derived and stored separately from `view_transform` and `projection` for better stability.
+    pub(crate) view_projection: Mat4,
+    /// Size of each shadow map texel in world units.
+    pub(crate) texel_size: f32,
+}
+
+pub fn update_directional_light_cascades(
+    directional_light_shadow_map: Res<DirectionalLightShadowMap>,
+    views: Query<(Entity, &GlobalTransform, &Projection, &Camera)>,
+    mut lights: Query<(
+        &GlobalTransform,
+        &DirectionalLight,
+        &CascadeShadowConfig,
+        &mut Cascades,
+    )>,
+) {
+    let views = views
+        .iter()
+        .filter_map(|x| match x {
+            // TODO: orthographic camera projection support.
```

Comment:
Agreed that it will be as large close as far, but it may still make sense to devote more pixels to closer distances than further ones w.r.t to quality bang for buck :)